Descope OIDC and SAML Integration Guide (#660)
ngrok/ngrok-docs: docs/integrations/descope/sso-oidc.mdx

---

> description: Use Descope OIDC to secure access to ngrok tunnels

Reviewer: ```suggestion
description: Use Descope OIDC to secure access to ngrok endpoints
```

---

# Descope SSO (OpenID Connect)

:::tip TL;DR

> To secure access to ngrok with Descope Single Sign-On using OpenID Connect:

Reviewer: ```suggestion
To secure access to your ngrok endpoints with Descope Single Sign-On (SSO) using OpenID Connect (OIDC):
```

1. [Configure Descope SSO](#configure-Descope)
1. [Configure ngrok](#configure-ngrok)
1. [Test access to ngrok with Descope SSO](#test-sso)

:::

> This article details how to configure Descope as the primary Identity Provider for ngrok tunnels.

Reviewer: ```suggestion
This article details how to configure Descope as the primary Identity Provider for your ngrok endpoints.
```

By integrating Descope SSO with ngrok, you can:

- **Restrict access to ngrok tunnels** only to users authenticated via Descope
- **Use Descope security policies and MFA authenticators**.
- **Use the Descope Console to facilitate access to ngrok apps**.

## Supported Features

The ngrok integration with Descope supports:

> - **SP-Initiated SSO**: In this mode, users access ngrok edges and tunnels and are redirected to Descope for authentication.

Reviewer: ```suggestion
- **SP-Initiated SSO**: In this mode, users access ngrok endpoints and are redirected to Descope for authentication.
```

## Requirements

> To configure ngrok tunnels with Descope, you must have:

Reviewer: ```suggestion
To configure ngrok endpoints with Descope, you must have:
```

> - an Descope account with administrative rights to a Descope Project

Reviewer: ```suggestion
- A Descope account with administrative rights to a Descope Project
```

> - an ngrok enterprise account with an access token or admin access to configure edges with OpenID Connect.

Reviewer: ```suggestion
- An ngrok enterprise account with an access token or admin access to configure edges with OpenID Connect.
```

### Create a Client Secret in Descope

1. Go to [Access Keys](https://app.descope.com/accesskeys) in the Descope Console.

1. Click **+ Access Key** in the top right corner.

1. In the popup window, enter `ngrok` in the **Name** field, change the **Expiration** time to whatever you prefer, and then click **Generate Key**.

1. Once the **Access Key** has been created, make a note of it for later.

## **Step 2**: Configure ngrok {#configure-ngrok}

ngrok can leverage Descope SSO in two ways:

- From the ngrok CLI (using the `--oidc` parameter)
- From the ngrok dashboard

### **Option 1**: ngrok CLI

> **Note:** For this tutorial, we assume you have an app running locally (i.e., on localhost:3000) with the ngrok client installed.

1. Launch a terminal

1. Enter the following command to launch an ngrok tunnel with Descope SSO:

   ```bash
   ngrok http 3000 --oidc=DESCOPE_ISSUER_URL \
   --oidc-client-id=DESCOPE_CLIENT_ID \
   --oidc-client-secret=DESCOPE_CLIENT_SECRET \
   ```

   **Note**: Replace the following with values:

>    - DESCOPE_ISSUER_URL: This is your Issuer URL that you copied from the **Applications** page earlier.

Reviewer: ```suggestion
   - `DESCOPE_ISSUER_URL`: This is your Issuer URL that you copied from the **Applications** page earlier.
```

>    - DESCOPE_CLIENT_ID: This is your Descope Project ID, which can be found under [Project Settings](https://app.descope.com/settings/project).

Reviewer: ```suggestion
   - `DESCOPE_CLIENT_ID`: This is your Descope Project ID, which can be found under [Project Settings](https://app.descope.com/settings/project).
```

>    - DESCOPE_CLIENT_SECRET: The is the Descope Access Key that you created from earlier.

Reviewer: ```suggestion
   - `DESCOPE_CLIENT_SECRET`: The is the Descope Access Key that you created from earlier.
```

1. Launch a tunnel connected to your Descope edge:

:::tip Note
For this step, we assume you have an app running locally (i.e. on localhost:3000) with the ngrok client installed.
:::

1.  Click **Start a tunnel**.

1.  Click the **copy icon** next to the tunnel command.
    ![tunnel config](img/descope-2.png)

1.  Launch a tunnel:

    - Launch a terminal.
    - Paste the command but replace `http://localhost:80` with your localhost app address (i.e., `http://localhost:3000`).
    - Click **Enter** and an ngrok tunnel associated with your edge configuration will launch.

1.  To confirm that the tunnel is connected to your edge:

    - Return to the ngrok dashboard
    - Close the **Start a tunnel** and the **Tunnel group** tabs
    - Refresh the test edge page. Under traffic, you will see the message _You have 1 tunnel online. Start additional tunnels to begin load balancing._
      ![tunnel confirmed](img/descope-3.png)

1.  In the test edge, copy the **endpoint URL**. (You use this URL to test the Descope Authentication)
    ![tunnel url](img/descope-4.png)

## Step 3: Test the integration {#test-sso}

1. In your browser, launch an incognito window.

> 1. Access your ngrok tunnel (i.e., `https://descope-sso-test.ngrok.io` or using the copied endpoint URL).

Reviewer: ```suggestion
1. Access your ngrok endpoint (i.e., `https://descope-sso-test.ngrok.io` or using the copied URL).
```
